File: netlify/functions/vector_store.py
from pinecone import Pinecone
from typing import List, Dict, Any, Optional
import logging
from config import Config
from embeddings import embedding_service

logger = logging.getLogger(__name__)

class VectorStore:
    def __init__(self):
        self.config = Config()
        self.pc = Pinecone(api_key=self.config.PINECONE_API_KEY)
        self.index = self.pc.Index(self.config.PINECONE_INDEX_NAME)
        
    def search(self, query: str, top_k: Optional[int] = None) -> List[str]:
        """Search for relevant context"""
        if top_k is None:
            top_k = self.config.TOP_K
            
        try:
            query_embedding = embedding_service.get_embedding(query)
            
            search_response = self.index.query(
                vector=query_embedding,
                top_k=top_k,
                include_metadata=True
            )
            
            contexts = []
            for match in search_response.matches:
                if match.score > self.config.SIMILARITY_THRESHOLD:
                    text = match.metadata.get('text', '')
                    if text:
                        contexts.append(text)
            
            return contexts
            
        except Exception as e:
            logger.error(f"Error searching Pinecone: {e}")
            return self._fallback_search(query)
    
    def upsert_documents(self, documents: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Upsert documents to Pinecone"""
        vectors = []
        
        for doc_idx, doc in enumerate(documents):
            text = doc.get('text', '')
            metadata = doc.get('metadata', {})
            
            if not text:
                logger.warning(f"Skipping document {doc_idx}: no text provided")
                continue
                
            # Chunk the text
            chunks = self._chunk_text(text)
            
            # Create vectors for each chunk
            for chunk_idx, chunk in enumerate(chunks):
                try:
                    embedding = embedding_service.get_embedding(chunk)
                    
                    vector_id = f"doc_{doc_idx}_chunk_{chunk_idx}_{hash(chunk)}"
                    
                    vectors.append({
                        'id': vector_id,
                        'values': embedding,
                        'metadata': {
                            **metadata,
                            'text': chunk,
                            'chunk_index': chunk_idx,
                            'total_chunks': len(chunks),
                            'document_index': doc_idx
                        }
                    })
                except Exception as e:
                    logger.error(f"Error processing chunk {chunk_idx} of document {doc_idx}: {e}")
        
        if not vectors:
            raise ValueError("No valid vectors generated")
        
        # Upsert in batches
        batch_size = 100
        upserted_count = 0
        
        for i in range(0, len(vectors), batch_size):
            batch = vectors[i:i + batch_size]
            try:
                self.index.upsert(vectors=batch)
                upserted_count += len(batch)
                logger.info(f"Upserted batch {i//batch_size + 1}, total: {upserted_count}")
            except Exception as e:
                logger.error(f"Error upserting batch {i//batch_size + 1}: {e}")
        
        return {
            'documents_processed': len(documents),
            'vectors_created': len(vectors),
            'vectors_upserted': upserted_count
        }
    
    def _chunk_text(self, text: str) -> List[str]:
        """Chunk text into smaller pieces"""
        sentences = [s.strip() for s in text.split('.') if s.strip()]
        chunks = []
        current_chunk = ''
        
        for sentence in sentences:
            if len(current_chunk) + len(sentence) > self.config.MAX_CHUNK_SIZE and current_chunk:
                chunks.append(current_chunk.strip())
                
                # Add overlap
                words = current_chunk.split()
                overlap_words = words[-self.config.CHUNK_OVERLAP//10:]
                current_chunk = ' '.join(overlap_words) + ' ' + sentence
            else:
                current_chunk += (' ' if current_chunk else '') + sentence
        
        if current_chunk.strip():
            chunks.append(current_chunk.strip())
            
        return chunks
    
    def _fallback_search(self, query: str) -> List[str]:
        """Fallback search using keyword matching"""
        fallback_knowledge = [
            "Suresh Beekhani is an experienced AI developer specializing in machine learning, deep learning, and AI system architecture.",
            "Suresh offers AI services including chatbot development, computer vision, NLP, and predictive modeling.",
            "He has worked on projects in healthcare, legal tech, e-commerce, and fintech industries.",
            "You can contact Suresh through his website contact form or LinkedIn."
        ]
        
        query_words = query.lower().split()
        query_words = [w for w in query_words if len(w) > 2]
        
        relevant = []
        for knowledge in fallback_knowledge:
            knowledge_lower = knowledge.lower()
            if any(word in knowledge_lower for word in query_words):
                relevant.append(knowledge)
        
        return relevant

# Global instance
vector_store = VectorStore()

File: netlify/functions/vector-store.js
import { Pinecone } from '@pinecone-database/pinecone';
import { config } from './config.js';
import { embeddingService } from './embeddings.js';
import { RetryHandler } from './retry-handler.js';

class VectorStore {
  constructor() {
    if (config.PINECONE_API_KEY) {
      this.pc = new Pinecone({
        apiKey: config.PINECONE_API_KEY,
      });
      // Access the 'pdf' namespace where the data is stored
      this.index = this.pc.index(config.PINECONE_INDEX_NAME).namespace('pdf');
    }
  }

  validateConnection() {
    if (!config.PINECONE_API_KEY) {
      console.error('Pinecone API key is missing');
      return false;
    }
    if (!this.index) {
      console.error('Pinecone index not initialized');
      return false;
    }
    return true;
  }

  async search(query, topK = null) {
    if (topK === null) {
      topK = config.TOP_K;
    }

    try {
      if (!this.validateConnection()) {
        console.warn('Pinecone not available, returning empty context');
        return [];
      }

      console.log('[Pinecone] Searching with query:', query.substring(0, 50));

      // Get query embedding
      const queryEmbedding = await embeddingService.getEmbedding(query);
      console.log('[Pinecone] Generated embedding, dimension:', queryEmbedding.length);

      if (queryEmbedding.length !== config.EMBEDDING_DIMENSION) {
        console.error(`Embedding dimension mismatch: expected ${config.EMBEDDING_DIMENSION}, got ${queryEmbedding.length}`);
        return [];
      }

      // Search with retry
      const searchResponse = await this.searchWithRetry(queryEmbedding, topK);
      console.log('[Pinecone] Search response - matches:', searchResponse.matches?.length || 0);

      // Filter and extract contexts
      const contexts = this.filterByRelevance(searchResponse.matches || []);
      console.log('[Pinecone] Filtered contexts:', contexts.length);

      if (contexts.length > 0) {
        console.log('[Pinecone] ✓ Found relevant context');
      } else {
        console.warn('[Pinecone] ✗ No relevant context found (threshold:', config.SIMILARITY_THRESHOLD, ')');
      }

      return contexts;

    } catch (error) {
      console.error('[Pinecone] Error searching:', error.message);
      return [];
    }
  }

  async searchWithRetry(queryEmbedding, topK) {
    const operation = async () => {
      return await this.index.query({
        vector: queryEmbedding,
        topK: topK,
        includeMetadata: true
      });
    };

    return await RetryHandler.withRetry(operation, {
      maxRetries: config.RETRY_MAX_ATTEMPTS || 2,     // ✅ use config value
      baseDelay: config.RETRY_BASE_DELAY || 300,
      maxDelay: config.RETRY_MAX_DELAY || 1000,
      timeout: config.REQUEST_TIMEOUT || 8000        // ✅ use config value
    });
  }

  filterByRelevance(matches) {
    const contexts = [];
    
    if (matches.length > 0) {
      console.log('[Pinecone] Match scores:', matches.map(m => m.score.toFixed(3)).join(', '));
    }
    
    for (const match of matches) {
      if (match.score > config.SIMILARITY_THRESHOLD) {
        let text = match.metadata?.text;
        if (text) {
          // ✅ Trim context length to avoid token explosion
          if (text.length > 2000) {
            text = text.substring(0, 2000) + '... [truncated]';
          }
          contexts.push(text);
        }
      }
    }
    return contexts;
  }

  async getIndexStats() {
    try {
      if (!this.validateConnection()) {
        throw new Error('Pinecone connection not available');
      }

      const stats = await this.pc.index(config.PINECONE_INDEX_NAME).describeIndexStats();
      return stats;
    } catch (error) {
      console.error('[Pinecone] Error getting index stats:', error.message);
      throw error;
    }
  }
}

export const vectorStore = new VectorStore();
export { VectorStore };
